Numerical_solution_of_SEIRD.Rmd

```diff
@@ -0,0 +1,356 @@
+solver with a fixed, small step size to that of the LSODA adaptive step size
+solver (comomodel's default choice for the solver algorithm).
+
+
+```{r, fig.width=7, fig.height=5}
+times_dense <- seq(0, 50, by=0.001)
+print("Timing for Euler (fixed step):")
+ptm <- proc.time()
+solution_dense <- run(model, times_dense, solve_method='euler')
+proc.time() - ptm
+
+times_daily <- seq(0, 50, by=1)
+print("Timing for LSODA (adaptive step):")
+ptm <- proc.time()
+solution_adapt <- run(model, times_daily, solve_method='lsoda')
+proc.time() - ptm
+
+df <- solution_dense$changes
+df <- as.data.frame(df)
+
+# Sum the deaths and incidences from the dense output to obtain daily values
+changes_dense <- data.frame(time=NULL, value=NULL, compartment=NULL)
+for (t in times_daily) {
+  cases = subset(df, t-1<df[,1] & df[,1]<=t)
+  incidence = subset(cases, cases[,3]=="Incidence")
+  death = subset(cases, cases[,3]=="Deaths")
+  changes_dense = rbind(changes_dense, 
+                        list(time=t,
+                             value=sum(incidence$value),
+                             compartment="Incidence",
+                             age_range="0-150"))
+  changes_dense = rbind(changes_dense,
+                        list(time=t,
+                             value=sum(death$value),
+                             compartment="Deaths",
+                             age_range="0-150"))
+}
+
+# Save the labels and combine for plotting
+changes_adapt <- solution_adapt$changes
+changes_dense$method <- "Euler, Δt=0.001 Day"
+changes_adapt$method <- "LSODA (adaptive step)"
+all_changes <- rbind(changes_dense, changes_adapt)
+
+
+ggplot(subset(all_changes, all_changes[,3]=="Incidence"), 
+       aes(x = time, y = value, fill = method)) +
+  geom_bar(stat="identity", position = position_dodge()) +
+  labs(title="Incidence")
+
+ggplot(subset(all_changes, all_changes[,3]=="Deaths"), 
+       aes(x = time, y = value, fill = method)) +
+  geom_bar(stat="identity", position = position_dodge()) +
+  labs(title="Deaths")
+
+```
+Both methods achieve the same accurate solution. However, for the LSODA solver,
+no fiddling with step sizes is necessary: we merely provide the time points at 
+which we want to access the solution (daily), and the solver handles the
+selection of the step sizes. We also note the significant speed advantage of
+the LSODA solver, which may be more than 50 times faster (the actual runtimes
+will be system dependent).
+
+Adaptive solvers are not infallible. On highly challenging differential
+equations, and when supplied with insufficient tolerances for the simulation or
+inference task at hand, they may be responsible for errors just as pernicious
+as those arising from fixed step solvers with too large step sizes. However,
+for the simulation of compartmental epidemiological models such as those
+covered by the comomodels package, we highly recommend the use of adaptive
+solvers such as LSODA for speed and accuracy.
```

Comment:
Maybe add something like this "which is the default choice of solver in the comomodels" at the end?